code-examples/low-latency-sessions/embedded/main.go

```diff
@@ -0,0 +1,137 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"net/http"
+	"time"
+
+	"github.com/dgraph-io/ristretto"
+	"github.com/gin-gonic/gin"
+
+	client "github.com/ory/kratos-client-go"
+)
+
+const (
+	cacheSize = 1000
+	// For local development, it's requires to use
+	// ory proxy to make cookies work fine.
+	// Replace it with http://projectId.projects.oryapis.com
+	// for production usage.
+	sdkURL = "http://localhost:4000/.ory"
+	// For local development, it's requires to use
+	// ory proxy to make cookies work fine.
+	// Replace it with http://projectId.projects.oryapis.com/ui/login
+	// for production usage.
+	loginURL = "http://localhost:4000/.ory/ui/login"
+)
+
+type kratosMiddleware struct {
+	client *client.APIClient
+	cache  *ristretto.Cache
+}
+
+func NewMiddleware() (*kratosMiddleware, error) {
+	configuration := client.NewConfiguration()
+	configuration.Servers = []client.ServerConfiguration{
+		{
+			URL: sdkURL,
+		},
+	}
+
+	cache, err := ristretto.NewCache(&ristretto.Config{
+		NumCounters: 1e7,     // number of keys to track frequency of (10M).
+		MaxCost:     1 << 30, // maximum cost of cache (1GB).
+		BufferItems: 64,      // number of keys per Get buffer.
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &kratosMiddleware{
+		client: client.NewAPIClient(configuration),
+		cache:  cache,
+	}, nil
+}
+func (k *kratosMiddleware) Session() gin.HandlerFunc {
+	return func(c *gin.Context) {
+		session, err := k.validateSession(c.Request)
+		if err != nil {
+			c.Redirect(http.StatusFound, loginURL)
+			return
+		}
+		if !*session.Active {
+			c.Redirect(http.StatusFound, "http://your_endpoint")
+			return
+		}
+		c.Next()
+	}
+}
+func (k *kratosMiddleware) cacheSession(key string, sess *client.Session) error {
+	// Calculate TTL. Do not cache sessions without TTL.
+	//
+	// Cached session should evict on sess.ExpiresAt
```

Comment:
evict? 
You mean the session is removed from the cache?

Reply:
Yep